xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderClientSslContextProvider.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.certprovider;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.envoyproxy.envoy.config.core.v3.Node;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CertificateValidationContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext.CombinedCertificateValidationContext;
+import io.grpc.netty.GrpcSslContexts;
+import io.grpc.xds.Bootstrapper.CertificateProviderInfo;
+import io.grpc.xds.EnvoyServerProtoData.UpstreamTlsContext;
+import io.grpc.xds.internal.sds.trust.SdsTrustManagerFactory;
+import io.netty.handler.ssl.SslContextBuilder;
+import java.security.cert.CertStoreException;
+import java.security.cert.X509Certificate;
+import java.util.Map;
+
+/** A client SslContext provider using CertificateProviderInstance to fetch secrets. */
+final class CertProviderClientSslContextProvider extends CertProviderSslContextProvider {
+
+  private CertProviderClientSslContextProvider(
+      Node node,
+      Map<String, CertificateProviderInfo> certProviders,
+      CommonTlsContext.CertificateProviderInstance certInstance,
+      CommonTlsContext.CertificateProviderInstance rootCertInstance,
+      CertificateValidationContext staticCertValidationContext,
+      UpstreamTlsContext upstreamTlsContext,
+      CertificateProviderStore certificateProviderStore) {
+    super(
+        node,
+        certProviders,
+        certInstance,
+        checkNotNull(rootCertInstance, "Client SSL requires rootCertInstance"),
+        staticCertValidationContext,
+        upstreamTlsContext,
+        certificateProviderStore);
+  }
+
+  @Override
+  public SslContextBuilder getSslContextBuilder(
+          CertificateValidationContext certificateValidationContextdationContext)
+      throws CertStoreException {
+    SslContextBuilder sslContextBuilder =
+        GrpcSslContexts.forClient()
+            .trustManager(
+                new SdsTrustManagerFactory(
+                    savedTrustedRoots.toArray(new X509Certificate[0]),
+                    certificateValidationContextdationContext));
+    if (isMtls()) {
+      sslContextBuilder.keyManager(savedKey, savedCertChain);
```

Comment:
Can `savedKey` be null? I think it's not required for client ssl context builder. Just not sure if it's true for MTLS.

Reply:
It can be proved that `savedKey` can't be null at this point:
- [`CertProviderSslContextProvider.updateSslContextWhenReady`](https://github.com/grpc/grpc-java/blob/2cf81bc5072fde5062d1738c2a2f336f602b389f/xds/src/main/java/io/grpc/xds/internal/certprovider/CertProviderSslContextProvider.java#L104) is where it all starts. And in case of `isMtls()` *and* `savedKey != null` it calls
- `DynamicSslContextProvider.updateSslContext` which calls `CertProviderClientSslContextProvider.getSslContextBuilder`  - this code.

In case this SslContextProvider is set up for mTLS (as dictated by xDS) then the SslContextBuilder is supplied with savedKey when available.

Reply:
So this method is only called by `DynamicSslContextProvider.updateSslContext` not at will (which should be mention in javadoc), so the signature here should be `protected` rather than `public`.

Reply:
Correct. In `DynamicSslContextProvider` it is defined as
```
 protected abstract SslContextBuilder getSslContextBuilder
```
But over here the visibility was increased to `public` (by mistake). I'll fix that.

Reply:
Changed to `protected final` like all the other subclasses.